### PR TITLE
Add implementations of equals()/hashCode() for views of THashMap

### PR DESCRIPTION
**Pull request by serejke**
`THashMap.keySet()` appeared to miss implementation of `equals()/hashCode()`.
I hit a bug comparing keys of two maps:
```
Map<K, V> one
Map<K, V> two
// This assertion fails if one or two is THashMap.
assertEquals(one.keySet(), two.keySet());
```

### CODE REVIEW

**Comment by develar** [comment]
I think it is time to stop using this outdated and not maintained library.
Numerous times this topic was raised and I think we should make decision.

**Comment by serejke** [comment]
I totally agree. It doesn't seem production-quality. Especially, the concern is the lack of full-covering tests.
But as soon as migration might be hard and lengthy (as I got from previous discussions), I believe, bug fixes still make sense.

File: core/src/main/java/gnu/trove/THashMap.java [hunk 1]
```diff
@@ -743,6 +743,32 @@ public V setValue(V o) {
...
+        public final boolean equals(Object o) {
+            if (o == this) {
+                return true;
+            }
+            if (o instanceof Map.Entry) {
+                Map.Entry<?, ?> otherEntry = (Map.Entry<?, ?>) o;
+                Object otherKey = otherEntry.getKey();
+                Object otherValue = otherEntry.getValue();
+                //noinspection unchecked
+                if (!(key == otherKey || key != null && _hashingStrategy.equals(key, (K) otherKey))) {
```
Comment (amaembo) [inline core/src/main/java/gnu/trove/THashMap.java]: This implementation violates the `Entry` spec. It's written in the spec how exactly keys and values should be compared:

> Compares the specified object with this entry for equality. Returns true if the given object is also a map entry and the two entries represent the same mapping. More formally, two entries e1 and e2 represent the same mapping if
> ```java
>               (e1.getKey()==null ?
>                e2.getKey()==null : e1.getKey().equals(e2.getKey()))  &&
>               (e1.getValue()==null ?
>                e2.getValue()==null : e1.getValue().equals(e2.getValue()))
> ```
> This ensures that the equals method works properly across different implementations of the Map.Entry interface.

Moreover, this `equals` implementation violates the general contract for `Object.equals`. If compared with `Map.Entry` that comes from another Map, the hashingStrategy of this Map will be used, so it's possible to violate the symmetry of equals (a.equals(b) != b.equals(a)). Is it really necessary to use hashingStrategy here? 

We can compare hashingStrategy with custom comparator of the `TreeMap`. The custom comparator may be inconsistent with `equals`. Yet, `TreeMap.Entry` doesn't use the comparator to compare entries. Instead, it fulfills the `equals` contract. Also, `TreeMap.EntrySet` implementation delegates its equals to `AbstractSet` like you are doing, and it looks like everybody is fine with this. I suggest to reuse that experience here as well: ignore hashing strategy when comparing views.

File: core/src/main/java/gnu/trove/THashMap.java [hunk 2]
```diff
@@ -743,6 +743,32 @@ public V setValue(V o) {
...
+                }
+                return val == otherValue || (val != null && val.equals(otherValue));
+            }
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            HashProcedure hashProcedure = new HashProcedure();
+            hashProcedure.execute(key, val);
```
Comment (amaembo) [inline core/src/main/java/gnu/trove/THashMap.java]: The `Map.Entry.hashCode` is also fully specified:

> Returns the hash code value for this map entry. The hash code of a map entry e is defined to be:
> ```java
>               (e.getKey()==null   ? 0 : e.getKey().hashCode()) ^
>               (e.getValue()==null ? 0 : e.getValue().hashCode())
> ```           
> This ensures that `e1.equals(e2)` implies that `e1.hashCode()==e2.hashCode()` for any two Entries `e1` and `e2`, as required by the general contract of `Object.hashCode`.

I don't think the `HashProcedure` will produce the same algorithm. Please write it according to the spec:

```java
int keyHash = (key==null ? 0 : key.hashCode());
int valueHash = (val==null ? 0 : val.hashCode());
return keyHash ^ valueHash;
```

File: core/src/test/java/gnu/trove/MapTest.java [hunk 3]
```diff
@@ -354,6 +355,31 @@ public boolean execute(Integer a, String b) {
         assertEquals((int) (map.capacity() * map._loadFactor), map._maxSize);
     }
 
+    public static void testTHashMapViewsEquality() {
+      THashMap<String, Integer> oneMap = new THashMap<String, Integer>();
+      THashMap<String, Integer> twoMap = new THashMap<String, Integer>();
```
Comment (amaembo) [inline core/src/test/java/gnu/trove/MapTest.java]: It would be good to compare with non-THashMap implementation (e.g. usual HashMap) in test as well.

**Review by amaembo — CHANGES_REQUESTED** [review]

**Comment by develar** [comment]
@serejke Where do you use it? Maybe you can use fastutil instead?

**Comment by serejke** [comment]
@develar THashMaps are widely used in IDE indexes. I wanted to compare two `Map<K, V>`-s produced by `DataIndexer` and caught this bug. In the new code I will consider using `fastutil`.

**Comment by serejke** [comment]
@amaembo Thank you for careful code review and pointing to Javadoc.

Indeed, we can assert that hashing strategy is "consistent with equals" (term from Javadoc) and compare keys using default `.equals()`.

All suggestions have been applied. New tests have been added.

On the way I caught one more bug: `THashMap.values()` used to inherit `MapBackedView` that was `Set`, which is incorrect because `.values()` must be a `Collection` (to allow value duplicates for different keys). Got rid of `MapBackedView` and inherited `AbstractSet` / `AbstractCollection` independently.

**Comment by develar** [comment]
fastutil available for use in platform. https://docs.google.com/document/d/1qJztRa3u2VKfzDH4jqFMtUAWAPr7rljKMg5PexuOwxg/edit

File: core/src/main/java/gnu/trove/THashMap.java [hunk 4]
```diff
@@ -593,108 +608,10 @@ protected K keyForEntry(Map.Entry<K,V> entry) {
...
-            for (Object element : collection) {
-                if (remove(element)) {
-                    changed = true;
-                }
-            }
-            return changed;
-        }
-
-        public void clear() {
-            THashMap.this.clear();
```
Comment (amaembo) [inline core/src/main/java/gnu/trove/THashMap.java:634]: I would suggest to implement `clear()` as well. `AbstractCollection.clear()` implementation is slow.

**Review by amaembo — APPROVED** [review]
Looks good in general. If you decide to implement clear(), additional review round may be skipped as it's pretty straightforward.

**Pull request closed without merging**